File: src/adoWiki.ts
/**
 * Read-only client for the team's Azure DevOps wiki.
 *
 * The Custom Team wiki (renweb / Custom Development / Custom-Team.wiki by
 * default) holds the team's standards: report-card variables, integration
 * setup guides, quoting and testing policy, SQL and ColdFusion snippets. A
 * case-scoped analysis pulls the few pages that match the case so the fix is
 * checked against how the team actually does things.
 *
 * AUTH: the user's existing Azure CLI login. `az account get-access-token` for
 * the Azure DevOps resource gives a short-lived bearer token; nothing is stored
 * on disk and the token is never logged or sent anywhere but dev.azure.com.
 * The az argv is fixed and app-authored — no user or case text reaches it.
 *
 * FAILURE IS NOT FATAL: every caller treats WikiUnavailable as "analyze without
 * wiki references" and records why as a caveat. A missing `az`, an expired
 * login or a network blip must never block the analysis itself.
 *
 * CACHE: `.analysis/wiki/` (git-ignored). The page tree and each page's content
 * are reused for 24h — the wiki changes slowly, and a case run shouldn't spend
 * its time re-downloading it.
 */

import { spawn } from "node:child_process";
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { join } from "node:path";

import { readEnvFile } from "./creatioClient.js";
import { ANALYSIS_DIR, writeAtomic } from "./workspace.js";

/** The Azure DevOps resource id — fixed for every ADO organization. */
const ADO_RESOURCE = "499b84ac-1321-427f-aa17-267ca6975798";

export const WIKI_DIR = join(ANALYSIS_DIR, "wiki");
const TREE_PATH = join(WIKI_DIR, "tree.json");
const PAGES_DIR = join(WIKI_DIR, "pages");

const CACHE_MS = 24 * 3_600_000;
const AZ_TIMEOUT_MS = 30_000;
const HTTP_TIMEOUT_MS = 20_000;

export type WikiFailure = "disabled" | "az-missing" | "not-logged-in" | "http";

export class WikiUnavailable extends Error {
  constructor(
    message: string,
    readonly reason: WikiFailure
  ) {
    super(message);
  }
}

export interface WikiConfig {
  org: string;
  project: string;
  wiki: string;
  enabled: boolean;
  maxPages: number;
}

export function wikiConfig(): WikiConfig {
  const env = readEnvFile();
  const n = parseInt((env.CREATIO_WIKI_MAX_PAGES || "").trim(), 10);
  return {
    org: (env.ADO_WIKI_ORG || "renweb").trim(),
    project: (env.ADO_WIKI_PROJECT || "Custom Development").trim(),
    wiki: (env.ADO_WIKI_ID || "Custom-Team.wiki").trim(),
    enabled: !/^(0|false|no|off)$/i.test((env.CREATIO_WIKI_ENABLED || "").trim()),
    maxPages: Number.isNaN(n) ? 4 : Math.max(1, Math.min(10, n)),
  };
}

function apiBase(c: WikiConfig): string {
  return (
    `https://dev.azure.com/${encodeURIComponent(c.org)}/${encodeURIComponent(c.project)}` +
    `/_apis/wiki/wikis/${encodeURIComponent(c.wiki)}/pages`
  );
}

/** Browser link for a page — what the UI and the reports show. */
export function pageUrl(c: WikiConfig, path: string, id?: number): string {
  const base =
    `https://dev.azure.com/${encodeURIComponent(c.org)}/${encodeURIComponent(c.project)}` +
    `/_wiki/wikis/${encodeURIComponent(c.wiki)}`;
  if (id) return `${base}/${id}`;
  return `${base}?pagePath=${encodeURIComponent(path)}`;
}

// ---------------------------------------------------------------------------
// Token
// ---------------------------------------------------------------------------

export type AzRunner = (args: string[]) => Promise<{ code: number; stdout: string; stderr: string }>;

/** Spawn the Azure CLI. `az` is a .cmd shim on Windows, which needs a shell. */
const defaultAzRunner: AzRunner = (args) =>
  new Promise((resolve) => {
    let stdout = "";
    let stderr = "";
    let child;
    if (args.some((a) => !/^[\w.:-]+$/.test(a))) {
      resolve({ code: 2, stdout: "", stderr: "Refusing an az argument with shell metacharacters." });
      return;
    }
    try {
      // On Windows the shim needs a shell. Node deprecates passing an args
      // array WITH a shell (it concatenates unescaped), so hand it one command
      // string instead — safe because every token is app-authored and fixed.
      const opts = { windowsHide: true, stdio: ["ignore", "pipe", "pipe"] as ["ignore", "pipe", "pipe"] };
      child =
        process.platform === "win32"
          ? spawn(["az", ...args].join(" "), { ...opts, shell: true })
          : spawn("az", args, opts);
    } catch (e) {
      resolve({ code: 127, stdout: "", stderr: e instanceof Error ? e.message : String(e) });
      return;
    }
    const timer = setTimeout(() => child.kill(), AZ_TIMEOUT_MS);
    child.stdout.on("data", (d) => (stdout += d));
    child.stderr.on("data", (d) => (stderr += d));
    child.on("error", (e) => {
      clearTimeout(timer);
      resolve({ code: 127, stdout, stderr: stderr || e.message });
    });
    child.on("close", (code) => {
      clearTimeout(timer);
      resolve({ code: code ?? 1, stdout, stderr });
    });
  });

let azRunner: AzRunner = defaultAzRunner;
let cached: { token: string; expires: number } | null = null;

/** Test seam: swap the az runner and drop the cached token. */
export function setAzRunner(r: AzRunner | null): void {
  azRunner = r || defaultAzRunner;
  cached = null;
}

/** Parse `az account get-access-token -o json` output. Exported for tests. */
export function parseTokenOutput(stdout: string): { token: string; expires: number } {
  let o: Record<string, unknown>;
  try {
    o = JSON.parse(stdout);
  } catch {
    throw new WikiUnavailable("The Azure CLI returned something that isn't a token.", "not-logged-in");
  }
  const token = typeof o.accessToken === "string" ? o.accessToken : "";
  if (!token) throw new WikiUnavailable("The Azure CLI returned no access token.", "not-logged-in");
  // `expires_on` (epoch seconds) is present on newer CLIs; `expiresOn` is a
  // local-time string on all of them.
  let expires = typeof o.expires_on === "number" ? o.expires_on * 1000 : NaN;
  if (Number.isNaN(expires) && typeof o.expiresOn === "string") expires = Date.parse(o.expiresOn);
  if (Number.isNaN(expires)) expires = Date.now() + 30 * 60_000;
  return { token, expires };
}

export async function getToken(): Promise<string> {
  if (cached && cached.expires - 5 * 60_000 > Date.now()) return cached.token;
  const r = await azRunner(["account", "get-access-token", "--resource", ADO_RESOURCE, "-o", "json"]);
  if (r.code !== 0) {
    const err = r.stderr || "";
    if (r.code === 127 || /not recognized|not found|ENOENT/i.test(err)) {
      throw new WikiUnavailable(
        "The Azure CLI (az) isn't installed, so the team wiki was skipped. Install it and run `az login`.",
        "az-missing"
      );
    }
    throw new WikiUnavailable(
      "The Azure CLI isn't logged in (or the login expired), so the team wiki was skipped. Run `az login`.",
      "not-logged-in"
    );
  }
  cached = parseTokenOutput(r.stdout);
  return cached.token;
}

// ---------------------------------------------------------------------------
// HTTP
// ---------------------------------------------------------------------------

async function adoGet(url: string): Promise<{ json: any; etag: string }> {
  const token = await getToken();
  const ctl = new AbortController();
  const timer = setTimeout(() => ctl.abort(), HTTP_TIMEOUT_MS);
  let res: Response;
  try {
    res = await fetch(url, {
      headers: { Authorization: `Bearer ${token}`, Accept: "application/json" },
      signal: ctl.signal,
    });
  } catch (e) {
    throw new WikiUnavailable(
      `The team wiki couldn't be reached (${e instanceof Error ? e.message : String(e)}).`,
      "http"
    );
  } finally {
    clearTimeout(timer);
  }
  if (res.status === 401 || res.status === 403) {
    cached = null;
    throw new WikiUnavailable(
      "Azure DevOps refused the Azure CLI login for the team wiki. Run `az login` with an account that can read it.",
      "not-logged-in"
    );
  }
  if (!res.ok) {
    throw new WikiUnavailable(`The team wiki returned HTTP ${res.status}.`, "http");
  }
  return { json: await res.json(), etag: res.headers.get("etag") || "" };
}

// ---------------------------------------------------------------------------
// Tree + pages
// ---------------------------------------------------------------------------

export interface WikiPageInfo {
  path: string;
  id?: number;
  /** Has sub-pages — a section rather than a leaf. */
  section: boolean;
}

interface TreeCache {
  fetchedAt: string;
  wiki: string;
  pages: WikiPageInfo[];
}

/** Flatten the ADO page tree, dropping the root and image folders. Exported for tests. */
export function flattenTree(node: any): WikiPageInfo[] {
  const out: WikiPageInfo[] = [];
  const walk = (n: any) => {
    if (!n || typeof n.path !== "string") return;
    if (/\/\.[^/]*$/.test(n.path) || n.path.includes("/.")) return; // .images, .attachments
    const subs: any[] = Array.isArray(n.subPages) ? n.subPages : [];
    if (n.path !== "/") out.push({ path: n.path, id: typeof n.id === "number" ? n.id : undefined, section: subs.length > 0 });
    for (const s of subs) walk(s);
  };
  walk(node);
  return out;
}

function fresh(iso: string | undefined): boolean {
  const t = Date.parse(iso || "");
  return !Number.isNaN(t) && Date.now() - t < CACHE_MS;
}

export async function getWikiTree(opts: { refresh?: boolean } = {}): Promise<WikiPageInfo[]> {
  const c = wikiConfig();
  if (!c.enabled) throw new WikiUnavailable("Team wiki lookup is turned off (CREATIO_WIKI_ENABLED).", "disabled");
  const key = `${c.org}/${c.project}/${c.wiki}`;
  if (!opts.refresh) {
    try {
      const doc = JSON.parse(readFileSync(TREE_PATH, "utf8")) as TreeCache;
      if (doc.wiki === key && fresh(doc.fetchedAt) && Array.isArray(doc.pages)) return doc.pages;
    } catch {
      /* no cache yet */
    }
  }
  const { json } = await adoGet(`${apiBase(c)}?path=/&recursionLevel=Full&api-version=7.1`);
  const pages = flattenTree(json);
  const doc: TreeCache = { fetchedAt: new Date().toISOString(), wiki: key, pages };
  writeAtomic(TREE_PATH, JSON.stringify(doc, null, 2));
  return pages;
}

export interface WikiPage {
  path: string;
  id?: number;
  url: string;
  content: string;
}

function pageCachePath(key: string): string {
  return join(PAGES_DIR, createHash("sha1").update(key).digest("hex").slice(0, 16) + ".json");
}

/**
 * Tidy page Markdown for a prompt: drop image embeds (the analysis can't see
 * them and they're pure noise), keep links and text.
 */
export function cleanPageContent(md: string): string {
  return String(md || "")
    .replace(/!\[[^\]]*\]\([^)]*\)/g, "")
    .replace(/<img\b[^>]*>/gi, "")
    .replace(/\n{3,}/g, "\n\n")
    .trim();
}

export async function getWikiPage(path: string, opts: { refresh?: boolean } = {}): Promise<WikiPage> {
  const c = wikiConfig();
  if (!c.enabled) throw new WikiUnavailable("Team wiki lookup is turned off (CREATIO_WIKI_ENABLED).", "disabled");
  const key = `${c.org}/${c.project}/${c.wiki}|${path}`;
  const file = pageCachePath(key);
  if (!opts.refresh) {
    try {
      const doc = JSON.parse(readFileSync(file, "utf8"));
      if (doc.key === key && fresh(doc.fetchedAt)) return doc.page as WikiPage;
    } catch {
      /* not cached */
    }
  }
  const { json } = await adoGet(
    `${apiBase(c)}?path=${encodeURIComponent(path)}&includeContent=true&api-version=7.1`
  );
  const id = typeof json.id === "number" ? json.id : undefined;
  const page: WikiPage = {
    path: typeof json.path === "string" ? json.path : path,
    id,
    url: pageUrl(c, path, id),
    content: cleanPageContent(json.content || ""),
  };
  writeAtomic(file, JSON.stringify({ key, fetchedAt: new Date().toISOString(), page }, null, 2));
  return page;
}

/** Settings-tab probe: can we log in and list the wiki? */
export async function testWiki(): Promise<{ ok: boolean; pages?: number; message: string; reason?: WikiFailure }> {
  try {
    const pages = await getWikiTree({ refresh: true });
    const c = wikiConfig();
    return { ok: true, pages: pages.length, message: `Connected to ${c.wiki} — ${pages.length} pages.` };
  } catch (e) {
    if (e instanceof WikiUnavailable) return { ok: false, message: e.message, reason: e.reason };
    return { ok: false, message: e instanceof Error ? e.message : String(e), reason: "http" };
  }
}

File: creatio_case_lookup/ado_wiki.py
"""Read-only client for the team's Azure DevOps wiki.

The Custom Team wiki (renweb / Custom Development / Custom-Team.wiki by
default) holds the team's standards: report-card variables, integration
setup guides, quoting and testing policy, SQL and ColdFusion snippets. A
case-scoped analysis pulls the few pages that match the case so the fix is
checked against how the team actually does things.

AUTH: the user's existing Azure CLI login. `az account get-access-token` for
the Azure DevOps resource gives a short-lived bearer token; nothing is stored
on disk and the token is never logged or sent anywhere but dev.azure.com.
The az argv is fixed and app-authored — no user or case text reaches it, and
it is exec'd directly (no shell).

FAILURE IS NOT FATAL: every caller treats WikiUnavailable as "analyze without
wiki references" and records why as a caveat. A missing `az`, an expired
login or a network blip must never block the analysis itself.

CACHE: `.analysis/wiki/` (git-ignored). The page tree and each page's content
are reused for 24h — the wiki changes slowly, and a case run shouldn't spend
its time re-downloading it. Same files as the TypeScript build, so the two
share a cache.
"""

from __future__ import annotations

import asyncio
import hashlib
import json
import re
import shutil
import subprocess
import sys
from pathlib import Path
from typing import Any, Awaitable, Callable
from urllib.parse import quote

import httpx

from .env import read_env_file
from .workspace import ANALYSIS_DIR, iso_now, js_parse_int, now_ms, parse_iso_ms, write_atomic

# The Azure DevOps resource id — fixed for every ADO organization.
ADO_RESOURCE = "499b84ac-1321-427f-aa17-267ca6975798"

WIKI_DIR = Path(ANALYSIS_DIR) / "wiki"
TREE_PATH = WIKI_DIR / "tree.json"
PAGES_DIR = WIKI_DIR / "pages"

CACHE_MS = 24 * 3_600_000
AZ_TIMEOUT_S = 30
HTTP_TIMEOUT_S = 20

# WikiFailure: "disabled" | "az-missing" | "not-logged-in" | "http"


class WikiUnavailable(Exception):
    def __init__(self, message: str, reason: str):
        super().__init__(message)
        self.reason = reason

    @property
    def message(self) -> str:
        return str(self)


def wiki_config() -> dict:
    """{org, project, wiki, enabled, maxPages} — read live from .env."""
    env = read_env_file()
    n = js_parse_int((env.get("CREATIO_WIKI_MAX_PAGES") or "").strip())
    return {
        "org": (env.get("ADO_WIKI_ORG") or "renweb").strip(),
        "project": (env.get("ADO_WIKI_PROJECT") or "Custom Development").strip(),
        "wiki": (env.get("ADO_WIKI_ID") or "Custom-Team.wiki").strip(),
        "enabled": not re.fullmatch(r"(0|false|no|off)", (env.get("CREATIO_WIKI_ENABLED") or "").strip(), re.IGNORECASE),
        "maxPages": 4 if n is None else max(1, min(10, n)),
    }


def _enc(s: str) -> str:
    """``encodeURIComponent``."""
    return quote(str(s), safe="-_.!~*'()")


def _api_base(c: dict) -> str:
    return f"https://dev.azure.com/{_enc(c['org'])}/{_enc(c['project'])}/_apis/wiki/wikis/{_enc(c['wiki'])}/pages"


def page_url(c: dict, path: str, id: int | None = None) -> str:
    """Browser link for a page — what the UI and the reports show."""
    base = f"https://dev.azure.com/{_enc(c['org'])}/{_enc(c['project'])}/_wiki/wikis/{_enc(c['wiki'])}"
    if id:
        return f"{base}/{id}"
    return f"{base}?pagePath={_enc(path)}"


# ---------------------------------------------------------------------------
# Token
# ---------------------------------------------------------------------------

# An AzRunner takes the argv (without "az") and returns {code, stdout, stderr}.
AzRunner = Callable[[list[str]], Awaitable[dict]]

_SAFE_ARG = re.compile(r"[\w.:-]+", re.ASCII)


async def _default_az_runner(args: list[str]) -> dict:
    """Run the Azure CLI with a fixed argv and no shell.

    `shutil.which("az")` resolves the `az.cmd` shim on Windows; CreateProcess
    runs a .cmd path directly. The metacharacter guard is kept anyway: cmd.exe
    still parses a batch file's arguments, so every token must be plain.
    """
    if any(not _SAFE_ARG.fullmatch(a) for a in args):
        return {"code": 2, "stdout": "", "stderr": "Refusing an az argument with shell metacharacters."}
    exe = shutil.which("az")
    if not exe:
        return {"code": 127, "stdout": "", "stderr": "spawn az ENOENT"}
    kwargs: dict[str, Any] = {}
    if sys.platform == "win32":
        kwargs["creationflags"] = subprocess.CREATE_NO_WINDOW  # windowsHide
    try:
        proc = await asyncio.create_subprocess_exec(
            exe,
            *args,
            stdin=asyncio.subprocess.DEVNULL,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            **kwargs,
        )
    except OSError as e:
        return {"code": 127, "stdout": "", "stderr": str(e)}
    try:
        out, err = await asyncio.wait_for(proc.communicate(), AZ_TIMEOUT_S)
    except asyncio.TimeoutError:
        try:
            proc.kill()
        except ProcessLookupError:
            pass
        await proc.wait()
        return {"code": 1, "stdout": "", "stderr": "az timed out"}
    return {
        "code": proc.returncode if proc.returncode is not None else 1,
        "stdout": out.decode("utf-8", errors="replace"),
        "stderr": err.decode("utf-8", errors="replace"),
    }


_az_runner: AzRunner = _default_az_runner
_cached: dict | None = None


def set_az_runner(r: AzRunner | None) -> None:
    """Test seam: swap the az runner and drop the cached token."""
    global _az_runner, _cached
    _az_runner = r or _default_az_runner
    _cached = None


def parse_token_output(stdout: str) -> dict:
    """Parse `az account get-access-token -o json` output → {token, expires (epoch ms)}.
    Exported for tests."""
    try:
        o = json.loads(stdout)
    except (ValueError, TypeError):
        raise WikiUnavailable("The Azure CLI returned something that isn't a token.", "not-logged-in") from None
    if not isinstance(o, dict):
        o = {}
    token = o.get("accessToken") if isinstance(o.get("accessToken"), str) else ""
    if not token:
        raise WikiUnavailable("The Azure CLI returned no access token.", "not-logged-in")
    # `expires_on` (epoch seconds) is present on newer CLIs; `expiresOn` is a
    # local-time string on all of them.
    eo = o.get("expires_on")
    expires: float | None = eo * 1000 if isinstance(eo, (int, float)) and not isinstance(eo, bool) else None
    if expires is None and isinstance(o.get("expiresOn"), str):
        expires = parse_iso_ms(o["expiresOn"])
    if expires is None:
        expires = now_ms() + 30 * 60_000
    return {"token": token, "expires": expires}


async def get_token() -> str:
    global _cached
    if _cached and _cached["expires"] - 5 * 60_000 > now_ms():
        return _cached["token"]
    r = await _az_runner(["account", "get-access-token", "--resource", ADO_RESOURCE, "-o", "json"])
    if r.get("code") != 0:
        err = r.get("stderr") or ""
        if r.get("code") == 127 or re.search(r"not recognized|not found|ENOENT", err, re.IGNORECASE):
            raise WikiUnavailable(
                "The Azure CLI (az) isn't installed, so the team wiki was skipped. Install it and run `az login`.",
                "az-missing",
            )
        raise WikiUnavailable(
            "The Azure CLI isn't logged in (or the login expired), so the team wiki was skipped. Run `az login`.",
            "not-logged-in",
        )
    _cached = parse_token_output(r.get("stdout") or "")
    return _cached["token"]


# ---------------------------------------------------------------------------
# HTTP
# ---------------------------------------------------------------------------


async def _ado_get(url: str, label: str = "team wiki") -> dict:
    """GET an Azure DevOps API URL → {json, etag}. `label` names the thing being
    read in error messages ("team wiki", "team skills repo")."""
    global _cached
    token = await get_token()
    try:
        async with httpx.AsyncClient(timeout=HTTP_TIMEOUT_S) as client:
            res = await client.get(url, headers={"Authorization": f"Bearer {token}", "Accept": "application/json"})
    except httpx.HTTPError as e:
        raise WikiUnavailable(f"The {label} couldn't be reached ({str(e) or type(e).__name__}).", "http") from None
    if res.status_code in (401, 403):
        _cached = None
        raise WikiUnavailable(
            f"Azure DevOps refused the Azure CLI login for the {label}. Run `az login` with an account that can read it.",
            "not-logged-in",
        )
    if not (200 <= res.status_code < 300):
        raise WikiUnavailable(f"The {label} returned HTTP {res.status_code}.", "http")
    return {"json": res.json(), "etag": res.headers.get("etag") or ""}


# ---------------------------------------------------------------------------
# Tree + pages
#
# WikiPageInfo: {path, id?, section} — `section` means it has sub-pages.
# WikiPage:     {path, id?, url, content}
# ---------------------------------------------------------------------------


def _is_num(v: Any) -> bool:
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def flatten_tree(node: Any) -> list[dict]:
    """Flatten the ADO page tree, dropping the root and image folders. Exported for tests."""
    out: list[dict] = []

    def walk(n: Any) -> None:
        if not isinstance(n, dict) or not isinstance(n.get("path"), str):
            return
        path = n["path"]
        if re.search(r"/\.[^/]*\Z", path) or "/." in path:
            return  # .images, .attachments
        subs = n.get("subPages") if isinstance(n.get("subPages"), list) else []
        if path != "/":
            row: dict = {"path": path}
            if _is_num(n.get("id")):
                row["id"] = n["id"]
            row["section"] = len(subs) > 0
            out.append(row)
        for s in subs:
            walk(s)

    walk(node)
    return out


def _fresh(iso: Any) -> bool:
    t = parse_iso_ms(iso or "")
    return t is not None and now_ms() - t < CACHE_MS


def _dump(obj: Any) -> str:
    return json.dumps(obj, indent=2, ensure_ascii=False)


async def get_wiki_tree(refresh: bool = False) -> list[dict]:
    c = wiki_config()
    if not c["enabled"]:
        raise WikiUnavailable("Team wiki lookup is turned off (CREATIO_WIKI_ENABLED).", "disabled")
    key = f"{c['org']}/{c['project']}/{c['wiki']}"
    if not refresh:
        try:
            with open(TREE_PATH, encoding="utf-8") as f:
                doc = json.load(f)
            if doc.get("wiki") == key and _fresh(doc.get("fetchedAt")) and isinstance(doc.get("pages"), list):
                return doc["pages"]
        except (OSError, ValueError, AttributeError):
            pass  # no cache yet
    got = await _ado_get(f"{_api_base(c)}?path=/&recursionLevel=Full&api-version=7.1")
    pages = flatten_tree(got["json"])
    doc = {"fetchedAt": iso_now(), "wiki": key, "pages": pages}
    write_atomic(TREE_PATH, _dump(doc))
    return pages


def _page_cache_path(key: str) -> Path:
    return Path(PAGES_DIR) / (hashlib.sha1(key.encode("utf-8")).hexdigest()[:16] + ".json")


def clean_page_content(md: str) -> str:
    """Tidy page Markdown for a prompt: drop image embeds (the analysis can't see
    them and they're pure noise), keep links and text."""
    s = str(md or "")
    s = re.sub(r"!\[[^\]]*\]\([^)]*\)", "", s)
    s = re.sub(r"<img\b[^>]*>", "", s, flags=re.IGNORECASE)
    s = re.sub(r"\n{3,}", "\n\n", s)
    return s.strip(" \t\n\x0b\x0c\r ﻿  　")


async def get_wiki_page(path: str, refresh: bool = False) -> dict:
    c = wiki_config()
    if not c["enabled"]:
        raise WikiUnavailable("Team wiki lookup is turned off (CREATIO_WIKI_ENABLED).", "disabled")
    key = f"{c['org']}/{c['project']}/{c['wiki']}|{path}"
    file = _page_cache_path(key)
    if not refresh:
        try:
            with open(file, encoding="utf-8") as f:
                doc = json.load(f)
            if doc.get("key") == key and _fresh(doc.get("fetchedAt")):
                return doc["page"]
        except (OSError, ValueError, AttributeError, KeyError):
            pass  # not cached
    got = await _ado_get(f"{_api_base(c)}?path={_enc(path)}&includeContent=true&api-version=7.1")
    j = got["json"] if isinstance(got["json"], dict) else {}
    page_id = j.get("id") if _is_num(j.get("id")) else None
    page: dict = {"path": j["path"] if isinstance(j.get("path"), str) else path}
    if page_id is not None:
        page["id"] = page_id
    page["url"] = page_url(c, path, page_id)
    page["content"] = clean_page_content(j.get("content") or "")
    write_atomic(file, _dump({"key": key, "fetchedAt": iso_now(), "page": page}))
    return page


async def test_wiki() -> dict:
    """Settings-tab probe: can we log in and list the wiki? → {ok, pages?, message, reason?}"""
    try:
        pages = await get_wiki_tree(refresh=True)
        c = wiki_config()
        return {"ok": True, "pages": len(pages), "message": f"Connected to {c['wiki']} — {len(pages)} pages."}
    except WikiUnavailable as e:
        return {"ok": False, "message": str(e), "reason": e.reason}
    except Exception as e:  # noqa: BLE001 — surfaced to the UI verbatim
        return {"ok": False, "message": str(e), "reason": "http"}


# pytest would otherwise collect `test_wiki` if this module were imported into a test namespace.
test_wiki.__test__ = False  # type: ignore[attr-defined]
